honeybadgermpc/hbavss.py

```diff
@@ -176,3 +181,152 @@ def _get_dealer_msg(self, value):
             v = None if values is None else values[i]
             avss_tasks[i] = asyncio.create_task(self.avss(k*avss_id+i, v, dealer_id))
         return await asyncio.gather(*avss_tasks)
+
+
+class HbAvssBatch(HbAvss):
+    async def _process_avss_msg(self, avss_id, dealer_id, dispersal_msg):
+        tag = f"{dealer_id}-{avss_id}-AVSS"
+        send, recv = self.get_send(tag), self.subscribe_recv(tag)
+
+        def multicast(msg):
+            for i in range(self.n):
+                send(i, msg)
+
+        commitments, ephemeral_public_key, encrypted_witnesses = loads(
+            dispersal_msg)
+
+        # all_encrypted_witnesses: n
+        shared_key = pow(ephemeral_public_key, self.private_key)
+
+        shares = [None] * (self.t+1)
+        witnesses = [None] * (self.t+1)
+        # Decrypt
+        for k in range(self.t+1):
+            shares[k], witnesses[k] = SymmetricCrypto.decrypt(
+                str(shared_key).encode("utf-8"),
+                encrypted_witnesses[self.my_id * (self.t + 1) + k])
+
+        # Don't know why this doesn't work
+        # # Decrypt
+        # for k in range(self.t+1):
+        #     shares[k], witnesses[k] = SymmetricCrypto.decrypt(
+        #         str(shared_key).encode("utf-8"), encrypted_witnesses[self.my_id][k])
+
+        # verify & send all ok
+        for i in range(self.t+1):
+            if self.poly_commit.verify_eval(
+                    commitments[i], self.my_id+1, shares[i], witnesses[i]):
+                multicast(HbAVSSMessageType.OK)
+            else:
+                # will be replaced by sending out IMPLICATE message later
+                # multicast(HbAVSSMessageType.IMPLICATE)
+                logging.error("PolyCommit verification failed.")
+                raise HoneyBadgerMPCError("PolyCommit verification failed.")
+
+        # Bracha-style agreement
+        oks_recvd = 0
+        readys_recvd = 0
+        # if 2t+1 OK or t+1 READY -> send all ready
+        while oks_recvd < 2 * self.t + 1 and readys_recvd < self.t + 1:
+            _, avss_msg = await recv()  # First value is the `sid`
+            if avss_msg == HbAVSSMessageType.OK:
+                oks_recvd += 1
+            elif avss_msg == HbAVSSMessageType.READY:
+                readys_recvd += 1
+
+        multicast(HbAVSSMessageType.READY)
+
+        # if 2t+1 ready -> output shares
+        while readys_recvd < 2 * self.t + 1:
+            _, avss_msg = await recv()  # First value is the `sid`
+            if avss_msg == HbAVSSMessageType.READY:
+                readys_recvd += 1
+
+        return shares
+
+    def _get_dealer_msg(self, value):
+        # Sample a random degree-(t,t) bivariate polynomial φ(·,·)
+        # such that each φ(0,k) = sk and φ(i,k) is Pi’s share of sk
+        phi = [None] * (self.t+1)
```

Comment:
is there an assumption here that `value` (shoudl be renamed `values`) contains exactly `t+1` elements? We should make this more general because the efficiency will be poor if it is only `t+1` at a time. As explained in the hbAVSS writeup it's necessary to amortize some of the work (like the reliable broadcast overhead) against somewhat larger batches

Reply:
And at a minimum, if it only works for `t+1` values, that should be checked rather than just omitting the additional values

Reply:
I followed the paper and I thought that is exactly t+1. I can definitely make that change.

Reply:
Resolved